client/state-db/src/pruning.rs

```diff
@@ -24,21 +24,25 @@
 //! the death list.
 //! The changes are journaled in the DB.
 
-use crate::{to_meta_key, CommitSet, Error, Hash, MetaDb};
+use crate::{noncanonical::LAST_CANONICAL, to_meta_key, CommitSet, Error, Hash, MetaDb};
 use codec::{Decode, Encode};
 use log::{trace, warn};
-use std::collections::{HashMap, HashSet, VecDeque};
+use std::{
+	cmp,
+	collections::{HashMap, HashSet, VecDeque},
+};
 
-const LAST_PRUNED: &[u8] = b"last_pruned";
+pub const LAST_PRUNED: &[u8] = b"last_pruned";
 const PRUNING_JOURNAL: &[u8] = b"pruning_journal";
+// Default pruning window size plus a magic number keep most common ops in cache
+const CACHE_BATCH_SIZE: usize = 256 + 10;
```

Comment:
We could expose 256 as another constant.
Also this size could be calculated by actual constraint value in state-db/src/lib.rs and not be constant.
Also, if the state-db do have a constraint > 256 (--pruning-size param), we could just disable the cache and always fetch from db on pop_front.
This is a alternative, just save some cache memory and avoid the batch download.

Reply:
> We could expose 256 as another constant.
> Also this size could be calculated by actual constraint value in state-db/src/lib.rs and not be constant.

Yes, we can calculate this cache size from the default window size, but overall we still need a constant to determine the max cache size. 

> Also, if the state-db do have a constraint > 256 (--pruning-size param), we could just disable the cache and always fetch from db on pop_front.
> This is a alternative, just save some cache memory and avoid the batch download.

Having a cache will be helpful. Since all blocks are eventually needed to prune (thus need to load from db if window size larger than the max cache size), batch loading doesn't bring more cost. 

And cache may be more friendly to the underlying caching (i.e page cache) as cache items are small (block hash + delete keys which is also hash I think) multiple items can fit in the same page cache. Also having a cache can avoid re-loading the block in case of `revert_pending` and it also makes the code clean as we don't need to add  
the `constraint > 256` condition around.

Reply:
> Having a cache will be helpful. Since all blocks are eventually needed to prune (thus need to load from db if window size larger than the max cache size), batch loading doesn't bring more cost.

The point of not having it is mainly to reduce code (maybe the query of 256 item would not be visible indeed).
As mention before, my opinion (may not be correct) on cache for pruning windows > cache size would be to use a sparse array (so not all item will have to be loaded/read from db). But really not a concern since first goal of this cache seems to be to maintain similar perf when using reasonable pruning window.